File: cpp/1574.shortest-subarray-to-be-removed-to-make-array-sorted.cpp
#include <algorithm>
#include <vector>
using std::vector;

class Solution {
public:
  int findLengthOfShortestSubarray(vector<int> &arr) {
    const int n = arr.size();
    int i = 0, j = n - 1;

    while (i < j && arr[i] <= arr[i + 1]) {
      i++;
    }

    if (i == n - 1) {
      return 0;
    }

    int minLen = n - i - 1;

    while (i < j && arr[j - 1] <= arr[j]) {
      j--;
    }

    minLen = std::min(minLen, j);

    int left = i;
    i = 0;

    while (i <= left && j < n) {
      if (arr[i] <= arr[j]) {
        minLen = std::min(minLen, j - i - 1);
        i++;
      } else {
        j++;
      }
    }

    return minLen;
  }
};

File: golang/1574.shortest-subarray-to-be-removed-to-make-array-sorted.go
func findLengthOfShortestSubarray(arr []int) int {
	n := len(arr)
	i, j := 0, n-1

	for i < j && arr[i] <= arr[i+1] {
		i++
	}

	if i == n-1 {
		return 0
	}

	minLen := n - i - 1

	for i < j && arr[j-1] <= arr[j] {
		j--
	}

	minLen = min(minLen, j)

	left := i
	i = 0

	for i <= left && j < n {
		if arr[i] <= arr[j] {
			minLen = min(minLen, j-i-1)
			i++
		} else {
			j++
		}
	}

	return minLen
}
